app/src/main/java/com/mapbox/maps/testapp/examples/SimulateNavigationRouteActivity.kt

```diff
@@ -0,0 +1,403 @@
+import com.mapbox.maps.plugin.animation.MapAnimationOptions
+import com.mapbox.maps.plugin.animation.camera
+import com.mapbox.maps.plugin.gestures.OnMapClickListener
+import com.mapbox.maps.plugin.gestures.OnMoveListener
+import com.mapbox.maps.plugin.gestures.gestures
+import com.mapbox.maps.plugin.locationcomponent.*
+import com.mapbox.maps.testapp.R
+import com.mapbox.maps.testapp.examples.annotation.AnnotationUtils
+import com.mapbox.turf.TurfMeasurement
+
+/**
+ * Simulate a navigation route with pre-defined route.
+ */
+class SimulateNavigationRouteActivity : AppCompatActivity() {
+
+  private lateinit var mapView: MapView
+  private lateinit var routePoints: LineString
+  private var renderFrameFinishCount = 0
+  private var overtimeFrameCount = 0
+  private val renderFrameIntervalsMs = mutableListOf<Float>()
+  private val locationProvider by lazy { FakeLocationProvider(routePoints) }
+  private var cameraFollowMode = CameraFollowMode.OVERVIEW
+  private val handler = Handler(Looper.getMainLooper())
+  private var startBenchmarkTime = 0L
+  private var lastRenderedFrameTime = 0L
+
+  private val onIndicatorBearingChangedListener = OnIndicatorBearingChangedListener {
+    if (cameraFollowMode == CameraFollowMode.FOLLOW) {
+      mapView.getMapboxMap()
+        .setCamera(CameraOptions.Builder().pitch(CAMERA_TRACKING_PITCH).bearing(it).build())
+    }
+  }
+
+  private val onIndicatorPositionChangedListener = OnIndicatorPositionChangedListener {
+    if (cameraFollowMode == CameraFollowMode.FOLLOW) {
+      mapView.getMapboxMap()
+        .setCamera(CameraOptions.Builder().center(it).pitch(CAMERA_TRACKING_PITCH).build())
+      mapView.gestures.focalPoint = mapView.getMapboxMap().pixelForCoordinate(it)
+    }
+  }
+
+  private val onMoveListener = object : OnMoveListener {
+    override fun onMoveBegin(detector: MoveGestureDetector) {
+      if (cameraFollowMode != CameraFollowMode.NONE) {
+        onCameraTrackingDismissed()
+      }
+    }
+
+    override fun onMove(detector: MoveGestureDetector): Boolean {
+      return false
+    }
+
+    override fun onMoveEnd(detector: MoveGestureDetector) {}
+  }
+
+  private val onMapClickListener = OnMapClickListener { point ->
+    when (cameraFollowMode) {
+      CameraFollowMode.OVERVIEW -> {
+        setCameraToFollowPuck()
+      }
+      CameraFollowMode.FOLLOW -> {
+        setCameraToOverview()
+      }
+      CameraFollowMode.NONE -> {
+        setCameraToFollowPuck()
+      }
+    }
+    true
+  }
+
+  private fun setCameraToOverview(easeTo: Boolean = true) {
+    cameraFollowMode = CameraFollowMode.OVERVIEW
+    val camera = mapView.getMapboxMap().cameraForCoordinates(
+      locationProvider.route.coordinates(),
+      bearing = 0.0,
+      pitch = 0.0,
+      padding = EdgeInsets(100.0, 100.0, 100.0, 100.0)
+    )
+    if (easeTo) {
+      mapView.camera.apply {
+        val zoom = createZoomAnimator(
+          cameraAnimatorOptions(camera.zoom!!)
```

Comment:
can be `camera.zoom ?: 0.0` ?

Reply:
the zoom in the camera options returned from the gl-native should never be null 😄 If it's null then we should throw exception.